lib/std/fmt.zig

```diff
@@ -625,30 +633,57 @@ pub fn formatAsciiChar(
     return writer.writeAll(@as(*const [1]u8, &c));
 }
 
+pub fn formatUnicodeCodepoint(
+    c: u21,
+    options: FormatOptions,
+    writer: anytype,
+) !void {
+    if (unicode.utf8ValidCodepoint(c)) {
+        var buf: [4]u8 = undefined;
+        // The codepoint is surely valid, hence the use of unreachable
+        const len = std.unicode.utf8Encode(@truncate(u21, c), &buf) catch |err| switch (err) {
+            error.Utf8CannotEncodeSurrogateHalf, error.CodepointTooLarge => unreachable,
+        };
+        return formatBuf(buf[0..len], options, writer);
+    }
+
+    // In case of error output the replacement char U+FFFD
+    return formatBuf(&[_]u8{ 0xef, 0xbf, 0xbd }, options, writer);
+}
+
 pub fn formatBuf(
     buf: []const u8,
     options: FormatOptions,
     writer: anytype,
 ) !void {
-    const width = options.width orelse buf.len;
-    const padding = if (width > buf.len) (width - buf.len) else 0;
-
-    switch (options.alignment) {
-        .Left => {
-            try writer.writeAll(buf);
-            try writer.writeByteNTimes(options.fill, padding);
-        },
-        .Center => {
-            const left_padding = padding / 2;
-            const right_padding = (padding + 1) / 2;
-            try writer.writeByteNTimes(options.fill, left_padding);
-            try writer.writeAll(buf);
-            try writer.writeByteNTimes(options.fill, right_padding);
-        },
-        .Right => {
-            try writer.writeByteNTimes(options.fill, padding);
-            try writer.writeAll(buf);
-        },
+    if (options.width) |min_width| {
+        // In case of error assume the buffer content is ASCII-encoded
+        const width = unicode.utf8CountCodepoints(buf) catch |_| buf.len;
```

Comment:
This won't decide the string width! A codepoint is not a single character. Example:
`"👩‍👦‍👦"` is U+1F469 U+200D U+1F466 U+200D U+1F466, which has 5 codepoints, but only width 1

You can look that up with this tool: https://cryptii.com/pipes/unicode-lookup

Reply:
> This won't decide the string width! 

That's a good approximation of the string width, the same approximation used by other PLs.
Entering the `wcwidth` territory and dealing with tables needing constant updates or mismatches between the producer (Zig, in this case) and the consumer (the terminal emulator/editor/browser) is definitely not something that I'd rank high on my todo list.

Reply:
Surely it would be nice if a user could put a table in the root source file and std.unicode apis could use it

Reply:
I think the best solution is to implement the runtime width specifier (see #1358) and let the user specify the display width, I'm playing with a prototype of this idea and it looks promising.